vendor/github.com/distribution/reference/normalize.go

```diff
@@ -56,41 +79,53 @@ func ParseNormalizedNamed(s string) (Named, error) {
-// mainly for backward compatibility.
-// The reference returned can only be either tagged or digested. For reference contains both tag
-// and digest, the function returns digested reference, e.g. docker.io/library/busybox:latest@
-// sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa will be returned as
-// docker.io/library/busybox@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa.
+// namedTaggedDigested is a reference that has both a tag and a digest.
+type namedTaggedDigested interface {
+	NamedTagged
+	Digested
+}
+
+// ParseDockerRef normalizes the image reference following the docker convention,
+// which allows for references to contain both a tag and a digest. It returns a
+// reference that is either tagged or digested. For references containing both
+// a tag and a digest, it returns a digested reference. For example, the following
+// reference:
+//
+//	docker.io/library/busybox:latest@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa
+//
+// Is returned as a digested reference (with the ":latest" tag removed):
+//
+//	docker.io/library/busybox@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa
+//
+// References that are already "tagged" or "digested" are returned unmodified:
+//
+//	// Already a digested reference
+//	docker.io/library/busybox@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa
+//
+//	// Already a named reference
+//	docker.io/library/busybox:latest
 func ParseDockerRef(ref string) (Named, error) {
 	named, err := ParseNormalizedNamed(ref)
 	if err != nil {
 		return nil, err
 	}
-	if _, ok := named.(NamedTagged); ok {
-		if canonical, ok := named.(Canonical); ok {
-			// The reference is both tagged and digested, only
-			// return digested.
-			newNamed, err := WithName(canonical.Name())
-			if err != nil {
-				return nil, err
-			}
-			newCanonical, err := WithDigest(newNamed, canonical.Digest())
-			if err != nil {
-				return nil, err
-			}
-			return newCanonical, nil
+	if canonical, ok := named.(namedTaggedDigested); ok {
+		// The reference is both tagged and digested; only return digested.
+		newNamed, err := WithName(canonical.Name())
+		if err != nil {
+			return nil, err
 		}
+		return WithDigest(newNamed, canonical.Digest())
 	}
 	return TagNameOnly(named), nil
 }
 
-// splitDockerDomain splits a repository name to domain and remotename string.
+// splitDockerDomain splits a repository name to domain and remote-name.
 // If no valid domain is found, the default domain is used. Repository name
 // needs to be already validated before.
 func splitDockerDomain(name string) (domain, remainder string) {
 	i := strings.IndexRune(name, '/')
-	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != localhost && strings.ToLower(name[:i]) == name[:i]) {
```

Comment:
the `strings.ToLower(name[:i]) == name[:i])` bit is interesting and was added in https://github.com/distribution/distribution/pull/2979/files

does this mean an image spec of `FOO/bar` that previously was treated as `docker.io/FOO/bar` is now treated as `FOO/bar`? does dockerhub disallow uppercase in repository names?

Reply:
@thaJeztah PTAL

Reply:
if dockerhub currently requires all namespaces to be lowercase, then this is fine, I just wanted to make sure no one was currently successfully using `Foo/bar` to refer to a `docker.io/Foo/bar` image

Reply:
ha, we got a unit test failure as well:

```
{Failed;  === RUN   TestParseImageName
    parsers_test.go:52: ParseImageName(ROOT) expects error couldn't parse image name "ROOT": invalid reference format: repository name (library/ROOT) must be lowercase but did not get one
--- FAIL: TestParseImageName (0.00s)
```

looks like we were leaning on the underlying parser to reject this name and it no longer does

Reply:
> if dockerhub currently requires all namespaces to be lowercase,

This is correct, Docker Hub does not allow mixed case, only lowercase, but it foes further than that, `docker` cli doesn't support uppercase repository names either. I am not familiar with other tools like `podman`, etc

Reply:
Thanks Milos! Yes, so it's definitely correct from that perspective; the only _possible_ change is that someone using `HELLO/world` would previously end up on `docker.io/HELLO/world` (and an error likely), and in the new situation, `HOST` would be considered a hostname, so potentially, a request can be made to `http(s)://HELLO/.....`

I don't expect huge issues with that, but it's worth mentioning.